### ISSUE TITLE
grub.cfg read as all 0's on ext2

### ISSUE THREAD

**Issue by eryjus** [jncronin/rpi-boot#19]
I have been working for the last few days to try to get rpi-boot to work and read my trivial kernel.  I think I have narrowed it down quite a bit and I'm hoping between my research and your knowledge we can get to the bottom of this.

First, this is a bit of log from my execution (your debugging statements with a few of mine thrown in as well):

> Performing read (size=35, block=8713)
> block_read: performing multi block read (1024 bytes; 2 blocks) from block 8713 on emmc0_0
> Calling device read at 0x22090
> mbr_read()
> CACHE: request for 1024 bytes starting at block 10761
> CACHE: request for buf_size 1024 - passing through to parent
> SD: issuing command CMD13
> DMA usage: off
> SD: command completed successfully
> SD: issuing command CMD12
> DMA usage: off
> SD: command completed successfully
> SD: issuing command CMD13
> DMA usage: off
> SD: command completed successfully
> SD: read() card ready, reading from block 10761
> SD: issuing command CMD18
> DMA usage: off
> SD: multi block transfer, awaiting block 0 ready
> 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000SD: block 0 transfer complete
> SD: multi block transfer, awaiting block 1 ready
> 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000SD: block 1 transfer complete
> SD: command completed successfully
> SD: data read successful
> SD: buffer is 
> Buffer is  (1024, 8713)
> The temporary buffer contains: ()
> The buffer now contains: ()
> MAIN: Parsing config file: ()
> Preparing to read first line...()

A couple of things to point out:
* grub.cfg is 35 bytes long and was properly located at block 8713
* The partition offset is properly calculated 2048 blocks out, resulting in a read of block 10761
* The list of 0's is the contents of each 4-byte read from the MMIO register
* All the previous reads work, meaning nothing else stands out at me as a problem

So, I am making my SD image with the following commands:
```
	mkdir -p ./p1
	dd if=/dev/zero of=rpi2b.img count=20 bs=1M
	parted --script rpi2b.img mklabel msdos mkpart p ext2 1 20 set 1 boot on
	sudo kpartx -as iso/rpi2b.img || true
	sudo mkfs.ext2 /dev/mapper/loop0p1
	sudo mount /dev/mapper/loop0p1 ./p1
	sudo cp -R sysroot/rpi2b/* p1/
	sudo umount ./p1
	sudo kpartx -d rpi2b.img
	rm -fR ./p1
```

Using `debugfs` on my SD image, I can see the following:
```
[adam@os-dev century]$ sudo kpartx -as rpi2b.img 
[adam@os-dev century]$ sudo debugfs /dev/mapper/loop0p1
debugfs 1.43.3 (04-Sep-2016)
debugfs:  bd 8713
0000  200a 6d75 6c74 6962 6f6f 7420 3d20 626f   .multiboot = bo
0020  6f74 2f6b 6572 6e65 6c2e 656c 660a 626f  ot/kernel.elf.bo
0040  6f74 0a00 0000 0000 0000 0000 0000 0000  ot..............
0060  0000 0000 0000 0000 0000 0000 0000 0000  ................
*
debugfs:  icheck 8713
Block   Inode number
8713    1541
debugfs:  ncheck 1541
Inode   Pathname
1541    /boot/grub/grub.cfg
```

Which tells me that the block in question really does hold the data I would like to read....

So, with all this, I was hoping you could point me to something else to review. Or perhaps this is a bug I need to open with QEMU?

In case you want to review the code I am working with (my kernel), you can find it here: https://github.com/eryjus/century

Thanks in advance!

**Comment by eryjus** [comment]
Update:

I am able to get my kernel to boot.  I had to use FAT16, rather than EXT2, in a similar situation to this other issue: https://github.com/jncronin/rpi-boot/issues/13

Based on what I am planning to support, I would like to support EXT2 rather than FAT.  Please let me know if there is anything you would like me to do or test.

**Comment by jncronin** [comment]
Hi,

Thanks for the report.  Could you confirm that offset 10761 * 512 bytes (=0x541200) in the image file contains the data you want.  If so, its a block read problem else its an ext2 problem.

Also, could you interrogate the ext2 filesystem somehow (either with debugging statements in rpi-boot or userspace tools) to ensure the ext2 block size is 512 bytes.

Thanks,
John.

**Comment by eryjus** [comment]
It took a little more research to get here, but I think I finally have a pattern.  My image changed and I neglected to save a copy of the original disk image (which I have now).

Here is what I have found out:
* Dumping the partition table with fdisk, the partition starts at sector 2048.
* Reviewing the ext2 superblock with debugfs, the block size is 1024.
* My new location for /boot/grub/grub.cfg is at block number 8705.

So, the math that actually located the data was:
(8705 * 1024) + (2048 * 512) = 0x980400
